README.md
# Holding Service

[![Build Status](https://travis-ci.com/NYPL-discovery/HoldingService.svg?branch=master)](https://travis-ci.com/NYPL-discovery/HoldingService) [![GitHub version](https://badge.fury.io/gh/nypl-discovery%2FHoldingService.svg)](https://badge.fury.io/gh/nypl-discovery%2FHoldingService)

## Purpose

This endpoint receives http requests containing new and updated holding statements, saves them, and passes them on to the configured Kinesis stream

## Running locally

```
rvm use
bundle install
psql -c 'create database test_holdings;' -U postgres
```

The method `init_db` in `holding-schema.rb` defines the schema for the `records` table, which the app will use. You can initialize this table by connecting to the db

```
ActiveRecord::Base.establish_connection(
  adapter: 'postgresql',
  database: ENV['DATABASE'],
  host: ENV['DB_HOST'],
  username: ENV['DB_USERNAME'],
  password: password
)
```

and running `init_db`


Lambdas with native extensions are a little tricky. You will need to bundle in the docker image:

Build the image:
```
docker build -t holding-service .
```

Enter the image:

```
docker run --rm -it -v $PWD:/var/task -w /var/task holding-service
```

In the image, install gems:

```
bundle install --path vendor/bundle --clean
```

You can then `exit` the image and proceed.

To test an event locally:

```
sam local invoke --region us-east-1 --template sam.local.yml --profile nypl-digital-dev --event events/empty.json
```

## Contributing

This repo follows a [PRS-Target-Master Git Workflow](https://github.com/NYPL/engineering-general/blob/a19c78b028148465139799f09732e7eb10115eef/standards/git-workflow.md#prs-target-master-merge-to-deployment-branches)

## Testing

Make sure you have bundle installed outside the docker image.

```
export SET BUNDLE_IGNORE_CONFIG=true; bundle exec rspec
```

## Deployment

Deployment is handled by travis for the following branches:

- `qa`
- `production`


## Managing

Credentials for the database are in the parameter store under `[env]/rds/holding_service/credentials`


CHANGELOG.md
# CHANGELOG

## 0.0.3 - 2021-07-26
### Added
- Added holdings?bib_ids= route to handle queries for multiple bib_ids
### Fix
- Updated query on `bibIds` column to use GIN index

## 0.0.2 - 2021-07-09
### Fix
- Fix holdings?bib_id= route https://github.com/NYPL-discovery/HoldingService/pull/33

## 0.0.1 - 2020-09-17
### Added
- Initial commits containing basic service for persisting Holding records


spec/db_init.rb
require 'pg'
require 'active_record'
require_relative '../models/record'
require_relative '../holding-schema'

def create_test_db
  ActiveRecord::Schema.define do
    drop_table :records, if_exists: true
  end

  init_db

  record_1 = {
    "id"=>1,
    "bibIds"=>[1, 2, 3],
    "bibIdLinks"=>[],
    "itemIds"=>[],
    "itemIdLinks"=>[],
    "inheritLocation"=>nil,
    "allocationRule"=>nil,
    "accountingUnit"=>1,
    "labelCode"=>"blah",
    "serialCode1"=>nil,
    "serialCode2"=>nil,
    "serialCode3"=>nil,
    "serialCode4"=>nil,
    "claimOnDate"=>nil,
    "receivingLocationCode"=>nil,
    "vendorCode"=>nil,
    "updateCount"=>nil,
    "pieceCount"=>nil,
    "eCheckInCode"=>nil,
    "mediaTypeCode"=>nil,
    "updatedDate"=>nil,
    "createdDate"=>"2020-07-29",
    "deletedDate"=>nil,
    "deleted"=>nil,
    "suppressed"=>nil,
    "fixedFields"=>{},
    "varFields"=>{},
    "holdings"=>{"a"=>1, "b"=>[2, 3]},
    "location"=>nil
  }

  record_2 = {
    "id"=>2,
    "bibIds"=>[3,4,5],
    "bibIdLinks"=>[],
    "itemIds"=>[],
    "itemIdLinks"=>[],
    "inheritLocation"=>nil,
    "allocationRule"=>nil,
    "accountingUnit"=>1,
    "labelCode"=>"blah",
    "serialCode1"=>nil,
    "serialCode2"=>nil,
    "serialCode3"=>nil,
    "serialCode4"=>nil,
    "claimOnDate"=>nil,
    "receivingLocationCode"=>nil,
    "vendorCode"=>nil,
    "updateCount"=>nil,
    "pieceCount"=>nil,
    "eCheckInCode"=>nil,
    "mediaTypeCode"=>nil,
    "updatedDate"=>nil,
    "createdDate"=>"2020-07-29",
    "deletedDate"=>nil,
    "deleted"=>nil,
    "suppressed"=>nil,
    "fixedFields"=>{},
    "varFields"=>{},
    "holdings"=>{"a"=>1, "b"=>[2, 3]},
    "location"=>nil
  }

  Record.create record_1
  Record.create record_2
end


spec/spec_helper.rb
require_relative 'db_init'

ENV['DATABASE']='test_holdings'
ENV['DB_HOST']='localhost'
ENV['DB_USERNAME']='postgres'
ENV['LOG_LEVEL']='info'

ActiveRecord::Base.establish_connection(
  adapter: 'postgresql',
  database: ENV['DATABASE'],
  host: ENV['DB_HOST'],
  username: ENV['DB_USERNAME']
)

create_test_db

class MockKinesisResponse
  def successful?
  end
end

class MockAwsKinesisClient
  def put_record(*args)
  end
end

# This file was generated by the `rspec --init` command. Conventionally, all
# specs live under a `spec` directory, which RSpec adds to the `$LOAD_PATH`.
# The generated `.rspec` file contains `--require spec_helper` which will cause
# this file to always be loaded, without a need to explicitly require it in any
# files.
#
# Given that it is always loaded, you are encouraged to keep this file as
# light-weight as possible. Requiring heavyweight dependencies from this file
# will add to the boot time of your test suite on EVERY test run, even for an
# individual file that may not need all of that loaded. Instead, consider making
# a separate helper file that requires the additional dependencies and performs
# the additional setup, and require it from the spec files that actually need
# it.
#
# See http://rubydoc.info/gems/rspec-core/RSpec/Core/Configuration
RSpec.configure do |config|
  # rspec-expectations config goes here. You can use an alternate
  # assertion/expectation library such as wrong or the stdlib/minitest
  # assertions if you prefer.
  config.expect_with :rspec do |expectations|
    # This option will default to `true` in RSpec 4. It makes the `description`
    # and `failure_message` of custom matchers include text for helper methods
    # defined using `chain`, e.g.:
    #     be_bigger_than(2).and_smaller_than(4).description
    #     # => "be bigger than 2 and smaller than 4"
    # ...rather than:
    #     # => "be bigger than 2"
    expectations.include_chain_clauses_in_custom_matcher_descriptions = true
  end

  # rspec-mocks config goes here. You can use an alternate test double
  # library (such as bogus or mocha) by changing the `mock_with` option here.
  config.mock_with :rspec do |mocks|
    # Prevents you from mocking or stubbing a method that does not exist on
    # a real object. This is generally recommended, and will default to
    # `true` in RSpec 4.
    mocks.verify_partial_doubles = true
  end

  # This option will default to `:apply_to_host_groups` in RSpec 4 (and will
  # have no way to turn it off -- the option exists only for backwards
  # compatibility in RSpec 3). It causes shared context metadata to be
  # inherited by the metadata hash of host groups and examples, rather than
  # triggering implicit auto-inclusion in groups with matching metadata.
  config.shared_context_metadata_behavior = :apply_to_host_groups

# The settings below are suggested to provide a good initial experience
# with RSpec, but feel free to customize to your heart's content.
=begin
  # This allows you to limit a spec run to individual examples or groups
  # you care about by tagging them with `:focus` metadata. When nothing
  # is tagged with `:focus`, all examples get run. RSpec also provides
  # aliases for `it`, `describe`, and `context` that include `:focus`
  # metadata: `fit`, `fdescribe` and `fcontext`, respectively.
  config.filter_run_when_matching :focus

  # Allows RSpec to persist some state between runs in order to support
  # the `--only-failures` and `--next-failure` CLI options. We recommend
  # you configure your source control system to ignore this file.
  config.example_status_persistence_file_path = "spec/examples.txt"

  # Limits the available syntax to the non-monkey patched syntax that is
  # recommended. For more details, see:
  #   - http://rspec.info/blog/2012/06/rspecs-new-expectation-syntax/
  #   - http://www.teaisaweso.me/blog/2013/05/27/rspecs-new-message-expectation-syntax/
  #   - http://rspec.info/blog/2014/05/notable-changes-in-rspec-3/#zero-monkey-patching-mode
  config.disable_monkey_patching!

  # This setting enables warnings. It's recommended, but in some cases may
  # be too noisy due to issues in dependencies.
  config.warnings = true

  # Many RSpec users commonly either run the entire suite or an individual
  # file, and it's useful to allow more verbose output when running an
  # individual spec file.
  if config.files_to_run.one?
    # Use the documentation formatter for detailed output,
    # unless a formatter has already been configured
    # (e.g. via a command-line flag).
    config.default_formatter = "doc"
  end

  # Print the 10 slowest examples and example groups at the
  # end of the spec run, to help surface which specs are running
  # particularly slow.
  config.profile_examples = 10

  # Run specs in random order to surface order dependencies. If you find an
  # order dependency and want to debug it, you can fix the order by providing
  # the seed, which is printed after each run.
  #     --seed 1234
  config.order = :random

  # Seed global randomization in this process using the `--seed` CLI option.
  # Setting this allows you to use `--seed` to deterministically reproduce
  # test failures related to randomization by passing the same `--seed` value
  # as the one that triggered the failure.
  Kernel.srand config.seed
=end
end


spec/app/app_spec.rb
require 'spec_helper'
require_relative '../../app.rb'
require_relative '../../src/http_methods.rb'

$mock_aws_kinesis = MockAwsKinesisClient.new

describe 'init' do
  before do
    allow(Aws::Kinesis::Client).to receive(:initialize).and_return($mock_aws_kinesis)
  end

  it 'should initialize db connection' do
    allow_any_instance_of(NYPLRubyUtil::KmsClient).to receive(:decrypt).and_return('')
    expect(ActiveRecord::Base).to receive(:establish_connection)
    init
  end

  it "should initialize logger" do
    expect($logger.class).to eq(NYPLRubyUtil::NyplLogFormatter)
  end

  it "should initialize db_fields" do
    expect($db_fields.class).to eq(Hash)
  end

  it "should initialize kinesis_client" do
    expect($kinesis_client.class).to eq(NYPLRubyUtil::KinesisClient)
  end

  it "should initialize swagger doc" do
    expect($swagger_doc.class).to eq(Hash)
  end

end

describe "handle_event" do

  swagger_event = {
    "path" => "/docs/holdings",
    "httpMethod" => "get"
  }

  mockMessage = "mockMessage"

  before do
    allow(Aws::Kinesis::Client).to receive(:new).and_return($mock_aws_kinesis)
    allow_any_instance_of(NYPLRubyUtil::KmsClient).to receive(:decrypt).and_return('')
    allow_any_instance_of(StandardError).to receive(:message).and_return(mockMessage)
    allow(self).to receive(:init).and_return(nil)
    allow($logger).to receive(:info).and_return(nil)
    allow($logger).to receive(:error).and_return(nil)
  end

  it "should call init" do
    expect(self).to receive(:init)
    handle_event(event: swagger_event, context: nil)
  end

  it "should log the event" do
    expect($logger).to receive(:info).with('handling event ', swagger_event)
    handle_event(event: swagger_event, context: nil)
  end

  it "should respond with swagger doc in case of swagger doc path" do
    resp = handle_event(event: swagger_event, context: nil)
    expect(resp).to eq(HTTPMethods.respond(200, $swagger_doc))
  end


  describe "get request" do
    get_event = {
      "path" => "",
      "httpMethod" => "GET"
    }

    mockMessage = 'mockMessage'

    before do
      allow(Aws::Kinesis::Client).to receive(:new).and_return($mock_aws_kinesis)
      allow_any_instance_of(NYPLRubyUtil::KmsClient).to receive(:decrypt).and_return('')
      allow_any_instance_of(StandardError).to receive(:message).and_return(mockMessage)
      allow(self).to receive(:init).and_return(nil)
      allow($logger).to receive(:info).and_return(nil)
      allow($logger).to receive(:error).and_return(nil)
    end

    it "should pass the event to get_holding" do
      expect(HTTPMethods).to receive(:get_holding).with(get_event)
      handle_event(event: get_event, context: nil)
    end
  end

  describe "post request" do
    post_event = {
      "path" => "",
      "httpMethod" => "POST"
    }

    mockMessage = 'mockMessage'

    before do
      allow(Aws::Kinesis::Client).to receive(:new).and_return($mock_aws_kinesis)
      allow_any_instance_of(NYPLRubyUtil::KmsClient).to receive(:decrypt).and_return('')
      allow_any_instance_of(StandardError).to receive(:message).and_return(mockMessage)
      allow(self).to receive(:init).and_return(nil)
      allow($logger).to receive(:info).and_return(nil)
      allow($logger).to receive(:error).and_return(nil)
    end

    it "should pass the event to post_holding" do
      expect(HTTPMethods).to receive(:post_holding).with(post_event)
      handle_event(event: post_event, context: nil)
    end
  end

  describe "post_holding" do
    mockMessage = 'mockMessage'

    before do
      allow(Aws::Kinesis::Client).to receive(:new).and_return($mock_aws_kinesis)
      allow_any_instance_of(NYPLRubyUtil::KmsClient).to receive(:decrypt).and_return('')
      allow_any_instance_of(StandardError).to receive(:message).and_return(mockMessage)
      allow(self).to receive(:init).and_return(nil)
      allow($logger).to receive(:info).and_return(nil)
      allow($logger).to receive(:error).and_return(nil)
    end

    malformed_json_event = {
      "path" => "",
      "httpMethod" => "",
      "body" => "fkajskajsd"
    }

    bad_record_event = {
      "path" => "",
      "httpMethod" => "",
      "body" => "[{}]"
    }

    correct_event = {
      "resource" => "/api/v0.1/holdings",
      "path" => "/api/v0.1/holdings",
      "httpMethod" => "POST",
      "body" => "[{\"id\":1140039,\"bibIds\":[17058685],\"bibIdLinks\":[\"https://nypl-sierra-test.nypl.org/iii/sierra-api/v6/bibs/17058685\"],\"itemIds\":[],\"itemIdLinks\":[],\"inheritLocation\":false,\"allocationRule\":null,\"accountingUnit\":1,\"labelCode\":\"n\",\"serialCode1\":\"a\",\"serialCode2\":\"-\",\"serialCode3\":\"-\",\"serialCode4\":\"-\",\"claimOnDate\":\"2020-06-07T04:00:00Z\",\"receivingLocationCode\":\"16\",\"vendorCode\":\"none\",\"updateCount\":\"-\",\"pieceCount\":0,\"eCheckInCode\":\" \",\"mediaTypeCode\":\" \",\"updatedDate\":\"2020-03-23T13:22:02Z\",\"createdDate\":\"2020-03-23T13:14:50Z\",\"deletedDate\":null,\"deleted\":false,\"suppressed\":false,\"fixedFields\":{\"35\":{\"label\":\"Label Type\",\"value\":\"n\",\"display\":null},\"36\":{\"label\":\"Serial Code 1\",\"value\":\"a\",\"display\":null},\"37\":{\"label\":\"Serial Code 2\",\"value\":\"-\",\"display\":null},\"38\":{\"label\":\"Copies\",\"value\":\"1\",\"display\":null},\"39\":{\"label\":\"Claim On\",\"value\":\"2020-06-07\",\"display\":null},\"40\":{\"label\":\"Location\",\"value\":\"mak  \",\"display\":null},\"41\":{\"label\":\"Recv Location\",\"value\":\"16 \",\"display\":null},\"42\":{\"label\":\"Vendor\",\"value\":\"none \",\"display\":null},\"80\":{\"label\":\"Record Type\",\"value\":\"c\",\"display\":null},\"81\":{\"label\":\"Record Number\",\"value\":\"1140039\",\"display\":null},\"83\":{\"label\":\"Created Date\",\"value\":\"2020-03-23T13:14:50Z\",\"display\":null},\"84\":{\"label\":\"Updated Date\",\"value\":\"2020-03-23T13:22:02Z\",\"display\":null},\"85\":{\"label\":\"No. of Revisions\",\"value\":\"2\",\"display\":null},\"86\":{\"label\":\"Agency\",\"value\":\"1\",\"display\":null},\"98\":{\"label\":\"PDATE\",\"value\":\"2020-03-23T13:14:00Z\",\"display\":null},\"118\":{\"label\":\"Serial Code 3\",\"value\":\"-\",\"display\":null},\"119\":{\"label\":\"Serial Code 4\",\"value\":\"-\",\"display\":null},\"120\":{\"label\":\"Update Count\",\"value\":\"-\",\"display\":null},\"121\":{\"label\":\"Piece Count\",\"value\":\"0\",\"display\":null},\"137\":{\"label\":\"E-Checkin\",\"value\":\" \",\"display\":null},\"159\":{\"label\":\"Media Type\",\"value\":\" \",\"display\":null},\"266\":{\"label\":\"Inherit Location\",\"value\":\"false\",\"display\":null}},\"varFields\":[{\"fieldTag\":\"_\",\"marcTag\":null,\"ind1\":null,\"ind2\":null,\"content\":\"00000ny   22000003n 4500\",\"subfields\":null},{\"fieldTag\":\"h\",\"marcTag\":\"863\",\"ind1\":\" \",\"ind2\":\" \",\"content\":null,\"subfields\":[{\"tag\":\"8\",\"content\":\"1.1\"},{\"tag\":\"a\",\"content\":\"1-3\"},{\"tag\":\"i\",\"content\":\"2020\"},{\"tag\":\"j\",\"content\":\"01-03\"},{\"tag\":\"k\",\"content\":\"15\"}]},{\"fieldTag\":\"i\",\"marcTag\":null,\"ind1\":null,\"ind2\":null,\"content\":\"TEST IDENTITY\",\"subfields\":null},{\"fieldTag\":\"n\",\"marcTag\":null,\"ind1\":null,\"ind2\":null,\"content\":\"Public note\",\"subfields\":null},{\"fieldTag\":\"y\",\"marcTag\":\"853\",\"ind1\":\" \",\"ind2\":\" \",\"content\":null,\"subfields\":[{\"tag\":\"8\",\"content\":\"1\"},{\"tag\":\"a\",\"content\":\"no.\"},{\"tag\":\"i\",\"content\":\"(year)\"},{\"tag\":\"j\",\"content\":\"(month)\"},{\"tag\":\"k\",\"content\":\"(day)\"}]},{\"fieldTag\":\"z\",\"marcTag\":null,\"ind1\":null,\"ind2\":null,\"content\":\"INTERNAL NOTE\",\"subfields\":null}],\"location\":{\"code\":\"mak\",\"label\":\"Schwarzman Building - Dewitt Wallace Room 108\"},\"holdings\":[]}]",
      "isBase64Encoded": false
    }

    updated_correct_event = {
      "resource" => "/api/v0.1/holdings",
      "path" => "/api/v0.1/holdings",
      "httpMethod" => "POST",
      "body" => "[{\"id\":1140039,\"bibIds\":[17058685],\"bibIdLinks\":[\"https://nypl-sierra-test.nypl.org/iii/sierra-api/v6/bibs/17058685\"],\"itemIds\":[],\"itemIdLinks\":[],\"inheritLocation\":true,\"allocationRule\":null,\"accountingUnit\":1,\"labelCode\":\"n\",\"serialCode1\":\"a\",\"serialCode2\":\"-\",\"serialCode3\":\"-\",\"serialCode4\":\"-\",\"claimOnDate\":\"2020-06-07T04:00:00Z\",\"receivingLocationCode\":\"16\",\"vendorCode\":\"none\",\"updateCount\":\"-\",\"pieceCount\":0,\"eCheckInCode\":\" \",\"mediaTypeCode\":\" \",\"updatedDate\":\"2020-03-23T13:22:02Z\",\"createdDate\":\"2020-03-23T13:14:50Z\",\"deletedDate\":null,\"deleted\":false,\"suppressed\":false,\"fixedFields\":{\"35\":{\"label\":\"Label Type\",\"value\":\"n\",\"display\":null},\"36\":{\"label\":\"Serial Code 1\",\"value\":\"a\",\"display\":null},\"37\":{\"label\":\"Serial Code 2\",\"value\":\"-\",\"display\":null},\"38\":{\"label\":\"Copies\",\"value\":\"1\",\"display\":null},\"39\":{\"label\":\"Claim On\",\"value\":\"2020-06-07\",\"display\":null},\"40\":{\"label\":\"Location\",\"value\":\"mak  \",\"display\":null},\"41\":{\"label\":\"Recv Location\",\"value\":\"16 \",\"display\":null},\"42\":{\"label\":\"Vendor\",\"value\":\"none \",\"display\":null},\"80\":{\"label\":\"Record Type\",\"value\":\"c\",\"display\":null},\"81\":{\"label\":\"Record Number\",\"value\":\"1140039\",\"display\":null},\"83\":{\"label\":\"Created Date\",\"value\":\"2020-03-23T13:14:50Z\",\"display\":null},\"84\":{\"label\":\"Updated Date\",\"value\":\"2020-03-23T13:22:02Z\",\"display\":null},\"85\":{\"label\":\"No. of Revisions\",\"value\":\"2\",\"display\":null},\"86\":{\"label\":\"Agency\",\"value\":\"1\",\"display\":null},\"98\":{\"label\":\"PDATE\",\"value\":\"2020-03-23T13:14:00Z\",\"display\":null},\"118\":{\"label\":\"Serial Code 3\",\"value\":\"-\",\"display\":null},\"119\":{\"label\":\"Serial Code 4\",\"value\":\"-\",\"display\":null},\"120\":{\"label\":\"Update Count\",\"value\":\"-\",\"display\":null},\"121\":{\"label\":\"Piece Count\",\"value\":\"0\",\"display\":null},\"137\":{\"label\":\"E-Checkin\",\"value\":\" \",\"display\":null},\"159\":{\"label\":\"Media Type\",\"value\":\" \",\"display\":null},\"266\":{\"label\":\"Inherit Location\",\"value\":\"false\",\"display\":null}},\"varFields\":[{\"fieldTag\":\"_\",\"marcTag\":null,\"ind1\":null,\"ind2\":null,\"content\":\"00000ny   22000003n 4500\",\"subfields\":null},{\"fieldTag\":\"h\",\"marcTag\":\"863\",\"ind1\":\" \",\"ind2\":\" \",\"content\":null,\"subfields\":[{\"tag\":\"8\",\"content\":\"1.1\"},{\"tag\":\"a\",\"content\":\"1-3\"},{\"tag\":\"i\",\"content\":\"2020\"},{\"tag\":\"j\",\"content\":\"01-03\"},{\"tag\":\"k\",\"content\":\"15\"}]},{\"fieldTag\":\"i\",\"marcTag\":null,\"ind1\":null,\"ind2\":null,\"content\":\"TEST IDENTITY\",\"subfields\":null},{\"fieldTag\":\"n\",\"marcTag\":null,\"ind1\":null,\"ind2\":null,\"content\":\"Public note\",\"subfields\":null},{\"fieldTag\":\"y\",\"marcTag\":\"853\",\"ind1\":\" \",\"ind2\":\" \",\"content\":null,\"subfields\":[{\"tag\":\"8\",\"content\":\"1\"},{\"tag\":\"a\",\"content\":\"no.\"},{\"tag\":\"i\",\"content\":\"(year)\"},{\"tag\":\"j\",\"content\":\"(month)\"},{\"tag\":\"k\",\"content\":\"(day)\"}]},{\"fieldTag\":\"z\",\"marcTag\":null,\"ind1\":null,\"ind2\":null,\"content\":\"INTERNAL NOTE\",\"subfields\":null}],\"location\":{\"code\":\"mak\",\"label\":\"Schwarzman Building - Dewitt Wallace Room 108\"},\"holdings\":[]}]",
      "isBase64Encoded": false
    }

    it "should log 'handling post request'" do
      expect($logger).to receive(:info).with("handling post request")
      HTTPMethods.post_holding(correct_event)
    end

    it "should log any problem parsing JSON" do
      expect($logger).to receive(:error).with('problem parsing JSON for event', { message: mockMessage })
      resp = HTTPMethods.post_holding(malformed_json_event)
    end

    it "should respond 500 in case of malformed json" do
      resp = HTTPMethods.post_holding malformed_json_event
      expect(resp).to eq(HTTPMethods.respond(500, { message: mockMessage }))
    end

    it "should log any error from saving record" do
      expect($logger).to receive(:error).with('problem persisting records to database', { message: mockMessage })
      resp = HTTPMethods.post_holding bad_record_event
    end

    it "should respond 500 in case of error saving record" do
      resp = HTTPMethods.post_holding bad_record_event
      expect(resp).to eq(HTTPMethods.respond(500, { message: mockMessage }))
    end

    it "should save the record in case of new record" do
      resp = HTTPMethods.post_holding correct_event
      expect(Record.find_by(id: 1140039)).to be_truthy
    end

    it "should update the record in case of old record" do
      resp = HTTPMethods.post_holding updated_correct_event
      expect(Record.find_by(id: 1140039).inherit_location).to eq(true)
    end

    it "should push record to kinesis" do
      allow($kinesis_client).to receive(:<<).and_return(nil)
      expect($kinesis_client).to receive(:<<).with(JSON.parse(correct_event["body"]).first)
      HTTPMethods.post_holding correct_event
    end

    it "should log in case of error in kinesis" do
      mock_response = MockKinesisResponse.new
      allow(mock_response).to receive(:successful?).and_return(false)
      allow($mock_aws_kinesis).to receive(:put_record).and_return(mock_response)
      expect($logger).to receive(:error)
      expect($mock_aws_kinesis).to receive(:put_record)
      HTTPMethods.post_holding correct_event
    end

    it "should respond 500 in case of error in kinesis" do
      mock_response = MockKinesisResponse.new
      allow(mock_response).to receive(:successful?).and_return(false)
      allow($mock_aws_kinesis).to receive(:put_record).and_return(mock_response)
      resp = HTTPMethods.post_holding correct_event
      expect(resp).to eq(HTTPMethods.respond(500, { message: mockMessage }))
    end

    it "should respond 200 in case run is successful" do
      mock_response = MockKinesisResponse.new
      allow(mock_response).to receive(:successful?).and_return(true)
      allow($mock_aws_kinesis).to receive(:put_record).and_return(mock_response)
      resp = HTTPMethods.post_holding correct_event
      expect(resp).to eq(HTTPMethods.respond(200, { "message": "Stored #{JSON.parse(correct_event["body"]).length} records Successfully", "errors": []}))
    end
  end
end

describe "get_holding" do
  before do
    allow($logger).to receive(:info).and_return(nil)
    allow($logger).to receive(:error).and_return(nil)
  end

  it "should log 'handling get request'" do
    get_event = {}
    expect($logger).to receive(:info).with('handling get request')
    HTTPMethods.get_holding(get_event)
  end

  it "should log params" do
    get_event = {'queryStringParameters' => { 'ids' => '1'}}
    expect($logger).to receive(:info)
    expect($logger).to receive(:info).with("params: #{get_event['queryStringParameters']}")
    HTTPMethods.get_holding(get_event)
  end

  describe "missing required params" do
    it "should respond 400 with the 'missing' message" do
      get_event = {}
      expect(HTTPMethods.get_holding(get_event)).to eq(HTTPMethods.respond(400, 'Must have bib_id or ids'))
    end
  end

  describe "both ids and bib_id set" do
    it "should respond 400 with error message" do
      get_event = {"queryStringParameters" => {"ids" => "1", "bib_id" => "2"}}
      expect(HTTPMethods.get_holding(get_event)).to eq(HTTPMethods.respond(400, "Can only have one of ids, bib_id"))
    end
  end

  describe "list of bib_ids" do
    it "should respond 400 with error message" do
      get_event = {"queryStringParameters" => {"bib_id" => "2,3"}}
      expect(HTTPMethods.get_holding(get_event)).to eq(HTTPMethods.respond(400, "bib_id must be a single numerical value"))
    end
  end

  describe "getting by id" do
    it "should log 'getting by ids'" do
      expect($logger).to receive(:info)
      expect($logger).to receive(:info)
      expect($logger).to receive(:info) do |params|
        expect(params).to include("getting by ids")
      end
      HTTPMethods.get_holding({'queryStringParameters' => { 'ids' => '1'}})
    end

    it "should return record for one id that exists" do
      event = {'queryStringParameters' => { 'ids' => '1'}}
      expect(HTTPMethods).to receive(:respond) do |*params|
        expect(params[0]).to eq(200)
        body = params[1]
        expect(body.length).to eq(1)
        expect(JSON.parse(body.first)["id"]).to eq(1)
      end
      HTTPMethods.get_holding(event)
    end

    it "should return records for two ids that exist" do
      event = {'queryStringParameters' => { 'ids' => '1,2'}}
      expect(HTTPMethods).to receive(:respond) do |*params|
        expect(params[0]).to eq(200)
        body = params[1]
        expect(body.length).to eq(2)
        expect(JSON.parse(body.first)["id"]).to eq(1)
        expect(JSON.parse(body.second)["id"]).to eq(2)
      end
      HTTPMethods.get_holding(event)
    end

    it "should return one record for one id that exists and one that doesn't" do
      event = {'queryStringParameters' => { 'ids' => '1,4'}}
      expect(HTTPMethods).to receive(:respond) do |*params|
        expect(params[0]).to eq(200)
        body = params[1]
        expect(body.length).to eq(1)
        expect(JSON.parse(body.first)["id"]).to eq(1)
      end
      HTTPMethods.get_holding(event)
    end

    it "should return empty array for one id that doesn't exist" do
      event = {'queryStringParameters' => { 'ids' => '4'}}
      expect(HTTPMethods).to receive(:respond) do |*params|
        expect(params[0]).to eq(200)
        body = params[1]
        expect(body.length).to eq(0)
      end
      HTTPMethods.get_holding(event)
    end

    it "should log responding 200" do
      event = {'queryStringParameters' => { 'ids' => '4'}}
      expect($logger).to receive(:info).with("responding 200")
      HTTPMethods.get_holding(event)
    end
  end

  describe "getting by bib_id" do
    it "should log 'getting by bib_id'" do
      expect($logger).to receive(:info)
      expect($logger).to receive(:info)
      expect($logger).to receive(:info) do |params|
        expect(params).to include("getting by bib_id")
      end
      HTTPMethods.get_holding({'queryStringParameters' => { 'bib_id' => '1'}})
    end

    it "should return record for one id that exists" do
      event = {'queryStringParameters' => { 'bib_id' => '1'}}
      expect(HTTPMethods).to receive(:respond) do |*params|
        expect(params[0]).to eq(200)
        body = params[1]
        expect(body.length).to eq(1)
        expect(JSON.parse(body.first)["id"]).to eq(1)
      end
      HTTPMethods.get_holding(event)
    end

    it "should return records for one id that exists multiple times" do
      event = {'queryStringParameters' => { 'bib_id' => '3'}}
      expect(HTTPMethods).to receive(:respond) do |*params|
        expect(params[0]).to eq(200)
        body = params[1]
        expect(body.length).to eq(2)
        expect(JSON.parse(body.first)["id"]).to eq(1)
        expect(JSON.parse(body.second)["id"]).to eq(2)
      end
      HTTPMethods.get_holding(event)
    end

    it "should return empty array for one id that doesn't exist" do
      event = {'queryStringParameters' => { 'bib_id' => '7'}}
      expect(HTTPMethods).to receive(:respond) do |*params|
        expect(params[0]).to eq(200)
        body = params[1]
        expect(body.length).to eq(0)
      end
      HTTPMethods.get_holding(event)
    end

    it "should log responding 200" do
      event = {'queryStringParameters' => { 'bib_id' => '4'}}
      expect($logger).to receive(:info).with("responding 200")
      HTTPMethods.get_holding(event)
    end
  end

  describe 'db error' do
    message = "problem getting records with ids: 1, message: Mock Error Message"
    it 'should log the params and error' do
      allow($logger).to receive(:info).with("responding 200").and_raise(StandardError.new("Mock Error Message"))
      expect($logger).to receive(:error).with(message)
      HTTPMethods.get_holding({'queryStringParameters' => {'ids' => '1'}})
    end

    it 'should respond 500 with error message' do
      allow($logger).to receive(:info).with("responding 200").and_raise(StandardError.new("Mock Error Message"))
      expect(HTTPMethods.get_holding({'queryStringParameters' => {'ids' => '1'}})).to eq(HTTPMethods.respond(500, message))
    end
  end
end


holding-schema.rb
require 'active_record'

def init_db
  ActiveRecord::Schema.define do
    create_table :records do |t|
      t.index :id, unique: true
      t.integer :bibIds, array: true, default: []
      t.string :bibIdLinks, array: true, default: []
      t.integer :itemIds, array: true, default: []
      t.string :itemIdLinks, array: true, default: []
      t.boolean :inheritLocation
      t.string :allocationRule
      t.integer :accountingUnit
      t.string :labelCode
      t.string :serialCode1
      t.string :serialCode2
      t.string :serialCode3
      t.string :serialCode4
      t.string :claimOnDate
      t.string :receivingLocationCode
      t.string :vendorCode
      t.string :updateCount
      t.integer :pieceCount
      t.string :eCheckInCode
      t.string :mediaTypeCode
      t.date :updatedDate
      t.date :createdDate
      t.date :deletedDate
      t.boolean :deleted
      t.boolean :suppressed
      t.jsonb :fixedFields, default: {}
      t.jsonb :varFields, default: {}
      t.jsonb :holdings, default: {}
      t.jsonb :location, default: {}
      t.jsonb :checkInCards, default: {}
    end
  end
end


app.rb
require 'parallel'
require 'pg'
require 'nypl_ruby_util'
require_relative 'models/record'
require_relative 'src/http_methods'

require_relative 'holding-schema'

def init
  $logger = NYPLRubyUtil::NyplLogFormatter.new(STDOUT, level: ENV['LOG_LEVEL'])
  $logger.info 'env: ', ENV.sort
  kms_client =  NYPLRubyUtil::KmsClient.new(ENV['KMS_OPTIONS'] ? JSON.parse(ENV['KMS_OPTIONS']) : {})
  password = kms_client.decrypt(ENV['DB_PASSWORD'])
  ActiveRecord::Base.establish_connection(
    adapter: 'postgresql',
    database: ENV['DATABASE'],
    host: ENV['DB_HOST'],
    username: ENV['DB_USERNAME'],
    password: password
  )

  $kinesis_client = NYPLRubyUtil::KinesisClient.new(
    schema_string: ENV['SCHEMA_STRING'],
    stream_name: ENV['STREAM_NAME']
  )
  $swagger_doc = JSON.load(File.read('./swagger.json'))
end

def handle_event(event:, context:)
  init
  path = event["path"]
  method = event["httpMethod"].downcase

  $logger.info('handling event ', event)

  if method == 'get' && path == "/docs/holdings"
    return HTTPMethods.respond 200, $swagger_doc
  elsif method == 'get'
    return HTTPMethods.get_holdings_req event
  elsif method == 'post'
    return HTTPMethods.post_holding event
  else
    return HTTPMethods.respond 500, "Path '#{path}' or Method '#{method}' not allowed"
  end

end
